torch/csrc/utils/future.h

```diff
@@ -64,10 +65,15 @@ class TORCH_API Future final {
     value_ = std::move(value);
     completed_ = true;
 
+
     // Move callbacks to a vector on the stack so we can access it without
     // holding a lock
     std::vector<Callback> cbs;
     cbs.swap(callbacks_);
+    // if recording, run end callbacks.
+    if (rf_) {
+      rf_->end();
+    }
```

Comment:
Can we do this outside the lock?

Reply:
I think so, I don't see any need to hold the lock when invoking these callbacks.